cupy/_core/_scalar.pyx

```diff
@@ -65,12 +65,37 @@ cdef object _numpy_float_ = numpy.float_
 cdef object _numpy_complex_ = numpy.complex_
 
 
-cpdef str get_typename(dtype):
+cpdef tuple get_typename_with_preamble(dtype):
```

Comment:
Would it be possible to separate this? the preamble applies to string data types only so maybe having a separate routine for the preamble that is only called when the any of the args is a string may be better and avoid changing a lot of code in kernel.pyx

Reply:
Thanks for the look, I am on a summit right now, so will dive back in more later.  The reason I did it this way was to avoid checking for string explicitly.  Also so that tagging it on for structured dtypes should not have to tuch `kernel.pyx` anymore than.  (And in theory even allowing custom ones as long as they can hook in here.)

OTOH, yeah I can probably avoid quite a bit of code churn by just special casing (I don't remember how much code churn was for this and how much for preserving the length information).

Reply:
Sorry for the late followup.  The changes in `kernel.pyx` are roughly split in half, I think:
* Threading this in is one change (mostly simple changes I think)
* The other part is making sure that we don't lose the string length information.

I could try to at least refactor that into two commits for clarity.  It still seems somewhat useful to have a single point where we go from Python to C, rather than do this and then check again if any of the types in the kernel need additional preamble information?
But, I agree that I mainly think that, because I was hoping to re-use this for structured dtypes and potentially broaden it up even to allow users to hook into it.